pkg/keycloak/keycloak.go

```diff
@@ -358,47 +355,67 @@ func (h *Handler) deleteKeycloak(kc *v1alpha1.Keycloak) error {
 }
 
 func (h *Handler) reconcileRealms(kc *v1alpha1.Keycloak, kcClient KeycloakInterface) error {
-	logrus.Debugf("reconcile realms, keycloak: %v", kc.Name)
-	realmPairsList := map[string]*KeycloakRealmPair{}
+	logrus.Infof("reconcile realms (%v)", kc.Name)
 
-	objRealms := kc.Spec.Realms
+	specRealms := kc.Spec.Realms
 	kcRealms, err := kcClient.ListRealms()
 	if err != nil {
 		return errors.Wrap(err, "error retrieving realms from keycloak")
 	}
 
-	logrus.Debugf("kc realms: %#v, total: %v", kcRealms, len(kcRealms))
-	logrus.Debugf("obj realms: %#v, total: %v", objRealms, len(objRealms))
-
 	kcRealmMap := map[string]*v1alpha1.KeycloakRealm{}
-	for i := 0; i < len(kcRealms); i++ {
-		logrus.Debugf("kc realm %+v", kcRealms[i])
+	for i := range kcRealms {
+		logrus.Debugf("kc realm %v", kcRealms[i].ID)
 		kcRealmMap[kcRealms[i].ID] = kcRealms[i]
 	}
 
-	for _, realm := range objRealms {
-		logrus.Debugf("obj realm %+v", realm)
-		realmPairsList[realm.ID] = &KeycloakRealmPair{
-			ObjRealm: &realm,
-			KcRealm:  kcRealmMap[realm.ID],
+	for i := range specRealms {
+		logrus.Debugf("spec realm %v", specRealms[i].ID)
+		err = h.reconcileRealm(kcRealmMap[specRealms[i].ID], &specRealms[i], kcClient)
+		//This should try and reconcile all realms rather throwing an error on the first failure
+		if err != nil {
+			return errors.Wrapf(err, "error reconciling realm: %v", specRealms[i].ID)
 		}
 	}
```

Comment:
Always use the index instead of value in these range routines, also deliberately not creating an additional variable to hold it since we don't want a copy past into reconcileRealm.